src/query.rs

```diff
@@ -0,0 +1,98 @@
+use pest::Parser;
+use std::fmt;
+
+use crate::index::{Index, get_packet_index};
+
+pub fn run_query(root: &str, query: String) -> Result<String, QueryError> {
+    let index = get_packet_index(root)
+        .unwrap_or_else(|error| {
+            panic!("Could not build outpack index from root at {}: {:?}",
+                   root, error);
+        });
+    let parsed: QueryNode = parse_query(&query)?;
+    eval_query(index, parsed)
+}
+
+#[derive(Parser)]
+#[grammar = "query.pest"]
+pub struct QueryParser;
+
+enum QueryNode<'a> {
+    Latest,
+    Lookup(&'a str)
+}
+
+#[derive(Debug, Clone)]
+pub enum QueryError {
+    ParseError(pest::error::Error<Rule>),
+    EvalError(String)
+}
+
+impl fmt::Display for QueryError {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+      match self {
+          QueryError::ParseError(err) => write!(f, "Failed to parse query\n{}", err),
+          QueryError::EvalError(msg) => write!(f, "Failed to evaluate query\n{}", msg)
+      }
+  }
+}
+
+fn parse_query(query: &str) -> Result<QueryNode, QueryError> {
+    match QueryParser::parse(Rule::query, query) {
+        Ok(mut pairs) => {
+            // Below never fails as query has been parsed and we know query rule can only have 1
+            // expr and its inner can only be length 1 also (either latest or a string)
+            let query = pairs.next().unwrap().into_inner().next().unwrap();
```

Comment:
I was a bit confused by the calls to `next` here. If there is only one each time, how about `peek` intead? (then also lose the `mut` on line 42)
```suggestion
            let query = pairs.peek().unwrap().into_inner().peek().unwrap();
```

Similarly on line 49

Reply:
Thanks! Fixed this up